cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -324,14 +326,17 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         ProgramOptions programOptions = ProgramOptions.fromNotification(notification, GSON);
         ProgramDescriptor programDescriptor =
           GSON.fromJson(properties.get(ProgramOptionConstants.PROGRAM_DESCRIPTOR), ProgramDescriptor.class);
+        ProgramRunStatus existingStatus = getProgramRunStatusFromRunRecord(appMetadataStore.getRun(programRunId));
```

Comment:
Were you pointing to reuse this? https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java#L233
Its only fetched when the notification was of type `PROGRAM_HEART_BEAT`.

Reply:
hum. that's true. I am thinking, is there a way to not to get the run record again, since all the `recordProgramXXX` methods does a read of the existing status internally. So if it can return it as well, then it saves one extra trip to the store to read the status.

Reply:
Added a new class `RunRecordDetailWithExistingStatus` to be returned by `recordProgramStop`.